docs/reference/scripting/using.asciidoc

```diff
@@ -1,425 +1,471 @@
-For most contexts, the default cache size is `100`. For ingest contexts, the
-default cache size is `200`.
+For example, the following statement creates two new reference types:
+`my_example1` and `my_example2`:
 
-NOTE: The size of scripts is limited to 65,535 bytes. This can be
-changed by setting `script.max_size_in_bytes` setting to increase that soft
-limit, but if scripts are really large then a
-<<modules-scripting-engine,native script engine>> should be considered.
+[source,painless]
+----
+Example my_example1 = new Example(); <1>
+Example my_example2 = my_example1; <2>
+my_example2.my_int = 4;
+----
+<1> `my_example1` creates a new instance of `Example`.
+<2> `my_example2` creates a new instance of `my_example1`.
 
-[[scripts-and-search-speed]]
-=== Scripts and search speed
+By creating a new instance of `Example`, `my_example1` references `Example`
+and inherits all of the data and logic that it contains.
 
-Scripts can't make use of {es}'s index structures or related optimizations. This
-can sometimes result in slower search speeds.
+Similarly, `my_example2` references `my_example1` and inherits everything it
+contains. Even though `my_example2` references `my_example1`, you can
+modify `my_example2` without impacting `my_example1`.
 
-If you often use scripts to transform indexed data, you can speed up search by
-making these changes during ingest instead. However, that often means slower
-index speeds.
+Looking back, `Example` indicates that `my_example.my_int = 1`. The last line
+in the previous statement modifies the value of that variable by setting the
+value of `my_int` to `4`. The `Example` reference type includes the following
+definitions:
 
-.*Example*
-[%collapsible]
-=====
-An index, `my_test_scores`, contains two `long` fields:
+[source,painless]
+----
+my_example.my_int = 1;
+my_example.my_double = 2;
+double my_add = my_example.add();
+----
 
-* `math_score`
-* `verbal_score`
+The `my_add` variable would return `3` because it adds the values of
+`my_example.my_int` (`1`) and `my_example.my_double` (`2`). However,
+`my_example2` would return `6`, because the value of `my_example2.my_int` is
+now `4` and the value of `my_example.my_double` is `2`.
 
-When running searches, users often use a script to sort results by the sum of
-these two field's values.
+By creating a new reference type, you can access existing objects and build
+logic on top of them without changing how they operate.
 
-[source,console]
-----
-GET /my_test_scores/_search
-{
-  "query": {
-    "term": {
-      "grad_year": "2099"
-    }
-  },
-  "sort": [
-    {
-      "_script": {
-        "type": "number",
-        "script": {
-          "source": "doc['math_score'].value + doc['verbal_score'].value"
-        },
-        "order": "desc"
-      }
-    }
-  ]
-}
-----
-// TEST[s/^/PUT my_test_scores\n/]
+[[script-type-dynamic]]
+==== Dynamic types
+A dynamic type can represent the value of any primitive type or reference type
```

Comment:
@jdconrad we need to come up with better language here.

Reply:
`def` types can point to any value.  The underlying value may have different types during script execution.
```
def abc = 1;
abc = "a string";
```

This provides flexibility, if you used strong types above, there would be a compile error.

```
int abc = 1;
abc = "a string"; // Compile error
```

However, if you know what type you'll be referring to, using an concrete type will be a slight speed up.

Reply:
Thanks for the clarification @stu-elastic. How about reworking this section like this?

> A dynamic type can represent any value by using a single type named `def`. A
> `def` type value mimics the behavior of whatever value it represents at
> runtime. The underlying value can have different types during script execution,
> which provides flexibility in your script.
> 
> For example, the following script includes a `def` type named `abc` that is
> implicitly defined as an `int` with a value of `1`. The second line sets the
> value of `abc` to a string value, which would result in a compilation error if
> `abc` was strictly defined as `int abc`.
> 
> ```
> def abc = 1;
> abc = "a string";
> ```
> 
> Because a `def` type can mimic multiple types, it always represents
> the last child descendant type value of any type value when evaluated. The
> default value for a newly-declared `def` type variable is `null`.
> 
> NOTE: Using the `def` type can have a slight impact on performance. Use only
> primitive types and reference types directly when performance is critical. If
> you know the type you're referring to, using a concrete type instead of a `def`
> type results in faster script execution.